### Equipo 5:

- **A01663977** - Ludovic Delot Bravo
- **A017716269** - Mónica Estrada Mondragón
- **A01784359** - Gonzalo González Méndez
- **A01783752** - Anakarenina Serrano Ibarra


## 1. Carga y preprocesamiento de datos

In [14]:
# Step 1: Installation of dependencies and version verification

import subprocess
import sys

# List of required libraries
libraries = [
    'pandas',
    'scikit-learn',
    'ace_tools'
]

# Function to install libraries if not already installed
def install_dependency(library):
    try:
        # Try to import the library
        __import__(library)
        print(f"{library} is already installed.")
    except ImportError:
        # If not installed, install it
        print(f"Installing {library}...")
        subprocess.check_call([sys.executable, "-m", "pip", "install", library])

# Install all the required libraries
for library in libraries:
    install_dependency(library)

# Verify the versions of the libraries
import pandas as pd
import sklearn

# Display installed library versions
print("\nInstalled versions:")
print(f"pandas: {pd.__version__}")
print(f"scikit-learn: {sklearn.__version__}")


# Instalar librerías necesarias si no están instaladas
import subprocess
import sys
import time
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, recall_score, precision_score, roc_auc_score, confusion_matrix, classification_report

# Lista de librerías necesarias
librerias = [
    'pandas', 
    'scikit-learn', 
    'matplotlib', 
    'seaborn',
    'ace_tools'
]

# Función para instalar librerías
def instalar_dependencia(libreria):
    try:
        __import__(libreria)
        print(f"{libreria} ya está instalada.")
    except ImportError:
        print(f"Instalando {libreria}...")
        subprocess.check_call([sys.executable, "-m", "pip", "install", libreria])

# Instalar las librerías necesarias
for libreria in librerias:
    instalar_dependencia(libreria)


pandas is already installed.
Installing scikit-learn...
Installing ace_tools...

Installed versions:
pandas: 2.2.3
scikit-learn: 1.6.1
pandas ya está instalada.
Instalando scikit-learn...
matplotlib ya está instalada.
seaborn ya está instalada.
Instalando ace_tools...


## 2. Carga y preprocesamiento de datos
### 2.1 Cargar los datos desde la URL

In [15]:
# Paso 2.1: Cargar los datos desde la URL
import pandas as pd

# Cargar el dataset desde la URL
url = "https://raw.githubusercontent.com/ArttapsTec/BasesDeDatosTEC/main/datos_default2.csv"
datos = pd.read_csv(url)

# Verificar las primeras filas del dataset
print("Primeras filas del dataset:")
print(datos.head())


Primeras filas del dataset:
    Age  Sex  Cuota  Nomina   Tasa  TDC  Antiguedad  BuenPago
0  37.0  1.0  240.0     0.0  165.0  0.0         1.0         0
1  54.0  1.0  166.0     0.0  118.0  1.0         0.0         1
2  72.0  1.0  123.0     1.0  130.0  0.0         1.5         1
3  51.0  1.0    0.0     0.0   60.0  0.0         0.0         1
4  65.0  1.0  235.0     0.0  120.0  1.0         1.5         1


### 2.2 Verificar y eliminar valores nulos

In [16]:
# Paso 2.2: Verificar valores nulos
print("\nValores nulos por columna:")
print(datos.isnull().sum())

# Eliminar filas con valores nulos
datos_limpios = datos.dropna()

# Verificar que no queden valores nulos
print("\nDespués de eliminar valores nulos:")
print(datos_limpios.isnull().sum())



Valores nulos por columna:
Age           40
Sex           40
Cuota         40
Nomina        40
Tasa          40
TDC           40
Antiguedad    40
BuenPago       0
dtype: int64

Después de eliminar valores nulos:
Age           0
Sex           0
Cuota         0
Nomina        0
Tasa          0
TDC           0
Antiguedad    0
BuenPago      0
dtype: int64


## 3. Preparación de características y variable objetivo

### 3.1 Separar las características (X) y la variable objetivo (y)

In [17]:
# Paso 3.1: Separar características (X) y la variable objetivo (y)
X = datos_limpios.drop(columns=['BuenPago'])  # Eliminamos la columna 'BuenPago' para las características
y = datos_limpios['BuenPago']  # 'BuenPago' es nuestra variable objetivo

# Verificar las primeras filas de X (características) y y (objetivo)
print("\nPrimeras filas de X (características):")
print(X.head())
print("\nPrimeras filas de y (variable objetivo):")
print(y.head())



Primeras filas de X (características):
    Age  Sex  Cuota  Nomina   Tasa  TDC  Antiguedad
0  37.0  1.0  240.0     0.0  165.0  0.0         1.0
1  54.0  1.0  166.0     0.0  118.0  1.0         0.0
2  72.0  1.0  123.0     1.0  130.0  0.0         1.5
3  51.0  1.0    0.0     0.0   60.0  0.0         0.0
4  65.0  1.0  235.0     0.0  120.0  1.0         1.5

Primeras filas de y (variable objetivo):
0    0
1    1
2    1
3    1
4    1
Name: BuenPago, dtype: int64


## 4. División de los datos en entrenamiento y prueba (70/30)

### 4.1 Dividir los datos usando train_test_split

In [18]:
# Paso 4.1: Dividir los datos en entrenamiento (70%) y prueba (30%)
from sklearn.model_selection import train_test_split

# Dividir los datos en entrenamiento y prueba
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=18)

# Verificar tamaños de los conjuntos
print(f"\nTamaño del conjunto de entrenamiento: {X_train.shape}")
print(f"Tamaño del conjunto de prueba: {X_test.shape}")



Tamaño del conjunto de entrenamiento: (385, 7)
Tamaño del conjunto de prueba: (165, 7)


# 5. Creación de modelos base de Random Forest

### 5.1 Crear los modelos de Random Forest con diferentes profundidades

In [19]:
# Paso 5.1: Crear los modelos de Random Forest
from sklearn.ensemble import RandomForestClassifier

# Definir los modelos con diferentes profundidades (max_depth)
rf1 = RandomForestClassifier(max_depth=2, n_estimators=20, random_state=18)
rf2 = RandomForestClassifier(max_depth=3, n_estimators=20, random_state=18)
rf3 = RandomForestClassifier(max_depth=4, n_estimators=20, random_state=18)
rf4 = RandomForestClassifier(max_depth=5, n_estimators=20, random_state=18)
rf5 = RandomForestClassifier(max_depth=6, n_estimators=20, random_state=18)
rf6 = RandomForestClassifier(max_depth=7, n_estimators=20, random_state=18)

# Guardar los modelos en una lista
modelos = [rf1, rf2, rf3, rf4, rf5, rf6]


# 6. Definición de la métrica personalizada

### 6.1 Crear la función para la métrica personalizada

In [20]:
# Paso 6.1: Crear la métrica personalizada para evaluar los modelos
from sklearn.metrics import confusion_matrix

def metrica_personalizada(y_true, y_pred, peso_tp=2, peso_tn=4, peso_fp=-1, peso_fn=3):
    """
    Calcula una métrica personalizada basada en la matriz de confusión:
    - Verdaderos Positivos (TP), Verdaderos Negativos (TN), Falsos Positivos (FP), Falsos Negativos (FN).
    """
    tn, fp, fn, tp = confusion_matrix(y_true, y_pred).ravel()
    
    # Cálculo de la métrica personalizada
    score = tp * peso_tp + tn * peso_tn + fp * peso_fp + fn * peso_fn
    return score


# 7. Entrenamiento y evaluación de los modelos

In [21]:
# Paso 7.1: Entrenar y evaluar todos los modelos, y preparar los datos para las gráficas
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.metrics import accuracy_score, recall_score, precision_score, roc_auc_score

# Lista para almacenar los resultados de todos los modelos
resultados_modelos = []

# Evaluar todos los modelos con un bucle
for i, modelo in enumerate(modelos, 1):
    print(f"\nEntrenando y evaluando el Modelo {i} (max_depth={modelo.max_depth})")
    
    # Entrenar el modelo con los datos de entrenamiento
    modelo.fit(X_train, y_train)
    
    # Realizar predicciones sobre el conjunto de prueba
    y_pred = modelo.predict(X_test)
    y_prob = modelo.predict_proba(X_test)[:, 1]  # Probabilidades para la clase 1
    
    # Calcular las métricas de desempeño
    accuracy = accuracy_score(y_test, y_pred)  # Precisión
    recall = recall_score(y_test, y_pred)  # Recall
    precision = precision_score(y_test, y_pred)  # Precisión
    roc_auc = roc_auc_score(y_test, y_prob)  # ROC-AUC
    
    # Calcular la métrica personalizada
    score_personalizado = metrica_personalizada(y_test, y_pred)
    
    # Almacenar los resultados del modelo en un diccionario
    resultados_modelos.append({
        'Modelo': f"RandomForest (max_depth={modelo.max_depth})",
        'Precisión': accuracy,
        'Recall': recall,
        'Precision': precision,
        'ROC-AUC': roc_auc,
        'Métrica Personalizada': score_personalizado
    })
    
    # Imprimir los resultados para cada modelo
    print(f"Modelo {i} - Precisión: {accuracy:.4f}")
    print(f"Modelo {i} - Recall: {recall:.4f}")
    print(f"Modelo {i} - Precision: {precision:.4f}")
    print(f"Modelo {i} - ROC-AUC: {roc_auc:.4f}")
    print(f"Modelo {i} - Métrica Personalizada: {score_personalizado:.4f}")



Entrenando y evaluando el Modelo 1 (max_depth=2)
Modelo 1 - Precisión: 0.8061
Modelo 1 - Recall: 0.7955
Modelo 1 - Precision: 0.8333
Modelo 1 - ROC-AUC: 0.8594
Modelo 1 - Métrica Personalizada: 432.0000

Entrenando y evaluando el Modelo 2 (max_depth=3)
Modelo 2 - Precisión: 0.8121
Modelo 2 - Recall: 0.7841
Modelo 2 - Precision: 0.8519
Modelo 2 - ROC-AUC: 0.8573
Modelo 2 - Métrica Personalizada: 443.0000

Entrenando y evaluando el Modelo 3 (max_depth=4)
Modelo 3 - Precisión: 0.8000
Modelo 3 - Recall: 0.7614
Modelo 3 - Precision: 0.8481
Modelo 3 - ROC-AUC: 0.8655
Modelo 3 - Métrica Personalizada: 445.0000

Entrenando y evaluando el Modelo 4 (max_depth=5)
Modelo 4 - Precisión: 0.8061
Modelo 4 - Recall: 0.7614
Modelo 4 - Precision: 0.8590
Modelo 4 - ROC-AUC: 0.8514
Modelo 4 - Métrica Personalizada: 450.0000

Entrenando y evaluando el Modelo 5 (max_depth=6)
Modelo 5 - Precisión: 0.7879
Modelo 5 - Recall: 0.7500
Modelo 5 - Precision: 0.8354
Modelo 5 - ROC-AUC: 0.8577
Modelo 5 - Métrica Pers

# 8. Almacenamiento y Visualización de Resultados

### 8.1 Guardar los resultados de los modelos en un DataFrame

In [22]:
# Paso 8.1: Guardar los resultados de todos los modelos en un DataFrame
import pandas as pd

# Convertir los resultados a un DataFrame
resultados_df = pd.DataFrame(resultados_modelos)

# Mostrar los resultados en una tabla
resultados_df
# También imprimimos los resultados en consola para revisión
print("\nResultados de todos los modelos:")
print(resultados_df)



Resultados de todos los modelos:
                       Modelo  Precisión    Recall  Precision   ROC-AUC  \
0  RandomForest (max_depth=2)   0.806061  0.795455   0.833333  0.859430   
1  RandomForest (max_depth=3)   0.812121  0.784091   0.851852  0.857290   
2  RandomForest (max_depth=4)   0.800000  0.761364   0.848101  0.865481   
3  RandomForest (max_depth=5)   0.806061  0.761364   0.858974  0.851387   
4  RandomForest (max_depth=6)   0.787879  0.750000   0.835443  0.857659   
5  RandomForest (max_depth=7)   0.775758  0.727273   0.831169  0.851978   

   Métrica Personalizada  
0                    432  
1                    443  
2                    445  
3                    450  
4                    441  
5                    443  


# 9. Graficar las Métricas Comparativas

### 9.1 Graficar las métricas de los modelos

In [23]:
import seaborn as sns
import matplotlib.pyplot as plt

sns.set(style="whitegrid", palette="muted")
plt.figure(figsize=(16, 12))

# Métricas que deseas graficar
metricas = [
    ('Precisión', 'Comparación de Precisión'),
    ('Recall', 'Comparación de Recall'),
    ('Precision', 'Comparación de Precision'),
    ('ROC-AUC', 'Comparación de ROC-AUC'),
    ('Métrica Personalizada', 'Comparación de Métrica Personalizada')
]

# Crear una gráfica por cada métrica
for i, (columna, titulo) in enumerate(metricas, start=1):
    plt.subplot(2, 3, i)
    ax = sns.barplot(x='Modelo', y=columna, data=resultados_df, palette='coolwarm')
    plt.title(titulo, fontsize=14)
    plt.xlabel('Modelo', fontsize=12)
    plt.ylabel(columna, fontsize=12)
    plt.xticks(rotation=45, ha='right')
    for p in ax.patches:
        ax.annotate(f'{p.get_height():.2f}', (p.get_x() + p.get_width() / 2., p.get_height()),
                    ha='center', va='center', fontsize=10, color='black',
                    xytext=(0, 5), textcoords='offset points')

plt.tight_layout()
plt.show()


/var/folders/33/y5pqbn8s76b1k47km33_cxc40000gn/T/ipykernel_86379/2684005643.py:19: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  ax = sns.barplot(x='Modelo', y=columna, data=resultados_df, palette='coolwarm')
/var/folders/33/y5pqbn8s76b1k47km33_cxc40000gn/T/ipykernel_86379/2684005643.py:19: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  ax = sns.barplot(x='Modelo', y=columna, data=resultados_df, palette='coolwarm')
/var/folders/33/y5pqbn8s76b1k47km33_cxc40000gn/T/ipykernel_86379/2684005643.py:19: FutureWarning: 

Passing `palette` without assigning `hue` is deprecated and will be removed in v0.14.0. Assign the `x` variable to `hue` and set `legend=False` for the same effect.

  ax = sns.barplot(x='Modelo', y=colum

# 10. Evaluación del mejor modelo

Consideramos que el modelo con max_depth de 5 es el mejor de todos los que probamos, sobre todo porque tuvo el mejor resultado en nuestra métrica personalizada, con una puntuación de 450, que es la más alta y esta métrica le da más importancia a los verdaderos negativos (TN) y penaliza los falsos positivos (FP) y falsos negativos (FN), lo que nos ayuda a medir de manera más precisa cómo el modelo está manejando ciertos errores y bueno además de esto, este modelo también tuvo buenos resultados en otras métricas como precisión y recall, lo que lo hace aún más confiable.

Y también obtuvo una precisión de 0.858974, lo que significa que predijo correctamente casi el 86% de las veces, lo cual está bastante bien. También tuvo un recall de 0.761364, lo que indica que fue capaz de identificar correctamente un buen porcentaje de los casos positivos, aunque no fue el mejor en esa métrica y bueno, en cuanto a la ROC-AUC, que nos dice qué tan bien el modelo puede distinguir entre las dos clases, consiguió 0.851387, lo cual también es bastante alto. Aunque el modelo con max_depth=4 tuvo un ROC-AUC ligeramente mejor, creemos que el modelo RandomForest (max_depth=5) es el más equilibrado en general, ya que destaca en nuestra métrica personalizada y también funciona bien en las otras métricas importantes.